### DISCUSSION TITLE
no socket implementation registered

### DISCUSSION THREAD

**Discussion by 3157696** [strongswan/strongswan#2443] [Q&A] [answered]
HELLO
I recently made a fresh install of openwrt on my route. 
openwrt : OpenWrt 23.05.4 

`opkg install  kmod-crypto-gcm  strongswan-minimal strongswan-swanctl kmod-crypto-gcm strongswan-mod-gmpdh strongswan-mod-curve25519 strongswan-mod-openssl strongswan-mod-pem strongswan-mod-curl 
`

I get an error when I start a connection：(The saddr does not contain a port Similar to [#2282](https://github.com/strongswan/strongswan/discussions/2282) )
But the router does not have charon-systemd ， there is no strongswan-starter package installed，and cannot delete any one

According to dependency，The installation package includes strongswan-mod-socket-default and strongswan-mod-stroke，

```
Thu Sep  5 05:02:29 2024 daemon.info ipsec: 06[CFG] vici initiate IKE_SA '2vps'
Thu Sep  5 05:02:29 2024 daemon.info ipsec: 08[IKE] initiating IKE_SA 2vps[4] to 1.2.3.4
Thu Sep  5 05:02:29 2024 daemon.info ipsec: 08[ENC] generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
Thu Sep  5 05:02:29 2024 daemon.info ipsec: 08[NET] sending packet: from 192.168.1.1 to 1.2.3.4[500] (296 bytes)
Thu Sep  5 05:02:29 2024 daemon.info ipsec: 03[NET] no socket implementation registered, sending failed
Thu Sep  5 05:02:33 2024 daemon.info ipsec: 11[IKE] retransmit 1 of request with message ID 0
Thu Sep  5 05:02:33 2024 daemon.info ipsec: 11[NET] sending packet: from 192.168.1.1 to 1.2.3.4[500] (296 bytes)
Thu Sep  5 05:02:33 2024 daemon.info ipsec: 03[NET] no socket implementation registered, sending failed
......
Thu Sep  5 05:05:14 2024 daemon.info ipsec: 07[IKE] giving up after 5 retransmits
Thu Sep  5 05:05:14 2024 daemon.info ipsec: 07[IKE] establishing IKE_SA failed, peer not responding

```

ps | grep charon
 1925 root      1828 S    /usr/lib/ipsec/starter --daemon charon --nofork
 **1953 root     11280 S    /usr/lib/ipsec/charon
 5369 root     11608 S    /usr/lib/ipsec/charon**
10947 root      1144 R    grep charon

How should I do it?
 Thank you

**Comment by tobiasbrunner** [comment]
You have two instances of charon running. Why?

**Reply by 3157696** [reply]
Ah ?

The is my swanctl.conf file
```
Deleted
```
Is there any problem here?

BTW
If interfaces_use = br-lan is configured in strongswan.conf, can passthrough-lan in swanctl.conf be removed?

**Reply by tobiasbrunner** [reply]
> The is my swanctl.conf file
> ...
> Is there any problem here?

Why would that have anything to do with how many instances of the IKE daemon you have running?

> If interfaces_use = br-lan is configured in strongswan.conf, can passthrough-lan in swanctl.conf be removed?

No, why would you think so?

**Reply by 3157696** [reply]
> > The is my swanctl.conf file
> > ...
> > Is there any problem here?
> 
> Why would that have anything to do with how many instances of the IKE daemon you have running?
> 
OO
It may be that the strongswan-ipsec package was also installed when strongswan-minimal was installed. But if strongswan-minimal is not installed, swanctl cannot be started. 

This is also one of the problems that bothers me
```
root@OpenWrt:~# ipsec stop
root@OpenWrt:~# ps | grep charon
 1954 root     11280 S    /usr/lib/ipsec/charon
 5660 root      1144 S    grep charon
root@OpenWrt:~#
```
> > If interfaces_use = br-lan is configured in strongswan.conf, can passthrough-lan in swanctl.conf be removed?
> 
> No, why would you think so?
It should be interfaces_use of charon.plugins.bypass-lan -> interfaces_use = br-lan

**Reply by tobiasbrunner** [reply, marked as answer]
Looking at the [OpenWrt package sources](https://github.com/openwrt/packages/tree/master/net/strongswan), I see that there is an `ipsec.init` and a `swanctl.init` script. The first starts `starter`, which then starts `charon`, the other starts `charon` directly. They are apparently installed by the `strongswan-ipsec` and `strongswan-swanctl` packages, respectively.

`strongswan-ipsec` seems to be a dependency of `strongswan-full` and `strongswan-isakmp`, neither `strongswan-default` nor `strongswan-minimal` depend on it. But weirdly, the latter has a dependency on the _stroke_ plugin (`strongswan-mod-stroke`), which seems to have a dependency on `strongswan-ipsec` (via `BuildPlugin` call).

So maybe use `strongswan-default` instead of `strongswan-minimal` and report this to the OpenWrt package maintainers (I guess they could remove `strongswan-mod-stroke` from `strongswan-minimal` and other packages and add that dependency to `strongswan-ipsec`, similar to how `strongswan-swanctl` depends on `strongswan-mod-vici`).

> It should be interfaces_use of charon.plugins.bypass-lan -> interfaces_use = br-lan

Ah, I see. Maybe, but since that plugin does not seem to get built on OpenWrt, it won't help you.
